calibration_residual_kernals.py

```diff
@@ -1,77 +1,98 @@
+                        phase_cals[stream_idx,frequency_indices[tune]] = residuals % (2*np.pi)
             
             delay_residual_map[ant] = residual_delays
-            phase_residual_map[ant] = residual_phases
+            phase_cal_map[ant] = phase_cals
 
-        return delay_residual_map, phase_residual_map
+        return delay_residual_map, phase_cal_map
 
-def calc_residuals_from_ifft(ant_to_gains, observation_frequencies):
+def calc_residuals_from_ifft(ant_to_gains, observation_frequencies, current_phase_cals):
     """
     Accept mapping of antenna to gains.
-    Returns ant to residual delay and ant to residual phase maps.
+    Returns ant to residual delay and ant to phase cal maps.
 
     Args:
-        ant_to_gains : {<ant> : [[complex(gains_pol0_tune0)], [complex(gains_pol1_tune0)], [complex(gains_pol0_tune1)], [complex(gains_pol1_tune1)]], ...}
-        observation_frequencies : list of dimension (n_tunings, nchans) in Hz
+        ant_to_gains : A dictionary mapping of antenna name to complex gain matrix of dimension (n_streams, n_chans).
+                    {<ant> : [[complex(gains_pol0_tune0)], [complex(gains_pol1_tune0)], [complex(gains_pol0_tune1)], [complex(gains_pol1_tune1)]], ...}
+        observation_frequencies : list of frequencies in Hz of dimension (n_tunings, nchans)
+        current_phase_cals : A mapping of antenna to phase coefficients on the F-Engine at present - a matrix of dimension (n_streams, n_chans).
+                    {<ant> : [[phase_cal_pol0_tune0], [phase_cal_pol1_tune0], [phase_cal_pol0_tune1], [phase_cal_pol1_tune1]], ...}
     Return:
-        delay_residual_map : {<ant> : [[residual_delay_pol0_tune0], [residual_delay_pol1_tune0], [residual_delay_pol0_tune1], [residual_delay_pol1_tune1]]}, ...} in nanoseconds
-        phase_residual_map : {<ant> : [[residual_phase_pol0_tune0],[residual_phase_pol1_tune0],
-                                        [residual_phase_pol0_tune1],[residual_phase_pol1_tune1]]}, ...} in radians
+        delay_residual_map : A mapping of antenna to delay residual values in nanoseconds of dimension (n_streams)
+                    {<ant> : [residual_delay_pol0_tune0, residual_delay_pol1_tune0, residual_delay_pol0_tune1, residual_delay_pol1_tune1]}, ...}
+        phase_cal_map : A mapping of antenna to phase calibration value in radians of dimension (n_streams, n_chans).
+                    {<ant> : [[phase_pol0_tune0],[phase_pol1_tune0],
+                                        [phase_pol0_tune1],[phase_pol1_tune1]]}, ...} in radians
     """
     delay_residual_map = {}
-    phase_residual_map = {}
+    phase_cal_map = {}
+
     for ant, gain_matrix in ant_to_gains.items():
         gain_matrix = np.array(gain_matrix, dtype = np.complex64)
+
+        #Subtract the last applied phases from the gain (incase incorrect)
+        current_phase_matrix = np.exp(1j * np.array(current_phase_cals[ant]))
+        new_gain_matrix = gain_matrix * current_phase_matrix
```

Comment:
This is essentially where the magic happens - apologies the changes don't make this so clear